### DISCUSSION TITLE
FFMPEG format define?

### DISCUSSION THREAD

**Discussion by Thumbserelly** [dlemstra/Magick.NET#913] [Help]
Hi just wondering if anyone knew how to set the format for when it calls ffmpeg from command line?
I am having trouble opening the ouput videos in Windows, and when using ffmpeg from command line myself i've noticed adding  format=yuv420p seems to work. I was hoping I could get magick.net to use this when doing Write on a MagickFormat.Mp4

**Comment by dlemstra** [comment]
Can you please provide a sample video that can be used to produce this issue?

**Reply by Thumbserelly** [reply]
```
       static void Main(string[] args)
        {
            MagickImage[] images = new MagickImage[4];
            for (int i = 0; i < images.Length; i++)
            {
                images[i] = new MagickImage(AppDomain.CurrentDomain.BaseDirectory + $"IMG_000{i + 1}.JPG");
                var text = new DrawableText(300, 200, (i + 1).ToString());
                var pointSize = new DrawableFontPointSize(60);
                MagickGeometry geom = new MagickGeometry();
                geom.Height = 600;
                geom.Width = 400;
                geom.FillArea = true;
                images[i].AnimationDelay = 250;
                images[i].Resize(geom);
                images[i].Crop(geom, Gravity.Center);
                images[i].Draw(pointSize, text);
                images[i].RePage();
            }

            var imageCollection = new MagickImageCollection(images);
            imageCollection.Write("output.mp4", MagickFormat.Mp4);

            string result = String.Empty;
            using (Process p = new Process())
            {
                p.StartInfo.UseShellExecute = false;
                p.StartInfo.CreateNoWindow = true;
                p.StartInfo.RedirectStandardOutput = true;
                p.StartInfo.FileName = "./ffmpeg.exe";
                p.StartInfo.Arguments = "-y -i ./output.mp4 -s 400x600 -vf format=yuv420p  ./output_fixed.mp4";
                p.Start();
                p.WaitForExit();
                result = p.StandardOutput.ReadToEnd();
            }
        }
```

Here's a sample program I made to show the issue I get. So output.mp4 wont run in windows, but If I rerun the command and force -vf format=yuv420p it fixes the file. I'll try to attach the 2 videos here.

https://user-images.githubusercontent.com/6605304/116278277-4200d200-a77e-11eb-9e34-d0d1a53e5066.mp4


https://user-images.githubusercontent.com/6605304/116278279-42996880-a77e-11eb-8e56-6e0ca996bf2c.mp4

**Reply by Thumbserelly** [reply]
> 
> 
> Can you please provide a sample video that can be used to produce this issue?

Do you get video cant be played because file is corrupt on vid1?

**Reply by dlemstra** [reply]
Ffmpeg is able to read the first video but it looks like Windows does not want to play it. There is currently no option to specify the pixel format but I am investigating if we could add this to ImageMagick.

**Reply by dlemstra** [reply]
In the next version of Magick.NET you will be able to specify the pixel format like this:

```C#
var defines = new VideoWriteDefines(MagickFormat.Mp4)
{
    PixelFormat = "yuv420p",
};

imageCollection.Write("output.mp4", defines);
```

**Reply by Thumbserelly** [reply]
> 
> 
> In the next version of Magick.NET you will be able to specify the pixel format like this:
> 
> ```cs
> var defines = new VideoWriteDefines(MagickFormat.Mp4)
> {
>     PixelFormat = "yuv420p",
> };
> 
> imageCollection.Write("output.mp4", defines);
> ```

awesome :)  Thank you

**Reply by dlemstra** [reply]
The new release with this feature has been published yesterday.

**Reply by Thumbserelly** [reply]
Hey, not sure if its worth opening a new discussion, but since im doing everything entirely in the program now I dont need to write out to file then read it back in
I have been trying to use ToByteArray

I've never used the auto git linking source thing for debugging, but I was calling ToByteArray and it just gets stuck, I even surrounded it with a try catch hoping it was just throwing an error but the program wouldnt come out.

```
        /// <summary>
        /// Converts this instance to a <see cref="byte"/> array.
        /// </summary>
        /// <param name="defines">The defines to set.</param>
        /// <returns>A <see cref="byte"/> array.</returns>
        /// <exception cref="MagickException">Thrown when an error is raised by ImageMagick.</exception>
        public byte[] ToByteArray(IWriteDefines defines)
        {
            SetDefines(defines);
            return ToByteArray(defines);
        }

```
is this calling itself over and over? 
It doesn't stack overflow so I'm not sure if the debuggers just going a bit weird.

**Reply by dlemstra** [reply]
That is indeed a bug that you just found. I did not test this with a `MagickImageCollection`. I only tested this with a `MagickImage`. It is really strange that this does not stack overflow because it should at some point?

**Reply by Thumbserelly** [reply]
I left it for around 5 mins thinking it might just be taking a while to write, but writing to file then reading the file to bytes was a lot faster so thought cant be that. I'm running a 64bit application, so maybe that's letting it go on for a long time? 

Other than that, the file it produces works for me now so thank you :)

**Reply by dlemstra** [reply]
Can you use writing to a file and reading it as temporary work around until I publish a new release? It might be a month before I willl normally publish a new release.

**Reply by Thumbserelly** [reply]
Yes that's fine, cheers :)

**Reply by Thumbserelly** [reply]
Sorry I know it was so long ago but.
When using the ToByteArray(defines) I end up getting an error about unable to read blob in temp files.
Ever seen this before?

"unable to read blob 'C:/Users/Me/AppData/Local/Temp/magick-mpk7QTziImUEQi8nIxwvJvRRPiB5e5FU0.pam': No such file or directory @ error/blob.c/FileToBlob/1438"

<!DOCTYPE html>
<HTML>
<head>
<title>Document</title></head>
<body>



  | Name | Value | Type
-- | -- | -- | --
  | StackTrace | "   at ImageMagick.MagickExceptionHelper.Check(IntPtr exception) in /_/src/Magick.NET/Exceptions/MagickExceptionHelper.cs:line 18\r\n   at ImageMagick.MagickImageCollection.NativeMagickImageCollection.WriteStream(IMagickImage image, IMagickSettings`1 settings, ReadWriteStreamDelegate writer, SeekStreamDelegate seeker, TellStreamDelegate teller, ReadWriteStreamDelegate reader) in /_/src/Magick.NET/Native/MagickImageCollection.cs:line 710\r\n   at ImageMagick.MagickImageCollection.Write(Stream stream) in /_/src/Magick.NET/MagickImageCollection.cs:line 1426\r\n   at ImageMagick.MagickImageCollection.ToByteArray() in /_/src/Magick.NET/MagickImageCollection.cs:line 1280\r\n   at ImageMagick.MagickImageCollection.ToByteArray(IWriteDefines defines) in /_/src/Magick.NET/MagickImageCollection.cs:line 1293\r\n   at Magick.MagickImageProcessor.CompositeImage() in D:\\Me\\Magick\\MagickImageProcessor.cs:line 164" | string

**Reply by dlemstra** [reply]
It looks like ImageMagick was unable to write the temporary file @Thumbserelly. Which version did you test this with?

**Reply by Thumbserelly** [reply]
@dlemstra  Core 7.0.1 and Q8-AnyCPU 7.24.1, ive upgraded to 8.0.1 for both and it has the same error

**Reply by dlemstra** [reply]
I suspect it is something in your setup. Is it possible to create a small repository on GitHub that can be used to reproduce this?

**Reply by Thumbserelly** [reply]
@dlemstra 
https://github.com/Thumbserelly/ConsoleApp1
ive put it on here, hopefully it runs for you
You'll need to get the ffmpeg.exe from somewhere, it was too big for me to upload it to github

**Reply by dlemstra** [reply]
Thanks for helping me out @Thumbserelly. With your repository I was able to track down the problem. This issue will be resolved in the next release.
